File: cdm-sources/cdm/margin/schedule/functions/IsIRSwapWithCallableBermudanRightToEnterExitSwaps.java
package cdm.margin.schedule.functions;

import cdm.product.qualification.functions.Qualify_BaseProduct_IRSwap;
import cdm.product.template.EarlyTerminationProvision;
import cdm.product.template.EconomicTerms;
import cdm.product.template.ExerciseTerms;
import cdm.product.template.OptionExerciseStyleEnum;
import cdm.product.template.OptionalEarlyTermination;
import cdm.product.template.TerminationProvision;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.expression.ComparisonResult;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(IsIRSwapWithCallableBermudanRightToEnterExitSwaps.IsIRSwapWithCallableBermudanRightToEnterExitSwapsDefault.class)
public abstract class IsIRSwapWithCallableBermudanRightToEnterExitSwaps implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected Qualify_BaseProduct_IRSwap qualify_BaseProduct_IRSwap;

	/**
	* @param economicTerms 
	* @return is_Product 
	*/
	public Boolean evaluate(EconomicTerms economicTerms) {
		Boolean is_Product = doEvaluate(economicTerms);
		
		return is_Product;
	}

	protected abstract Boolean doEvaluate(EconomicTerms economicTerms);

	public static class IsIRSwapWithCallableBermudanRightToEnterExitSwapsDefault extends IsIRSwapWithCallableBermudanRightToEnterExitSwaps {
		@Override
		protected Boolean doEvaluate(EconomicTerms economicTerms) {
			Boolean is_Product = null;
			return assignOutput(is_Product, economicTerms);
		}
		
		protected Boolean assignOutput(Boolean is_Product, EconomicTerms economicTerms) {
			is_Product = ComparisonResult.of(MapperS.of(qualify_BaseProduct_IRSwap.evaluate(economicTerms))).and(areEqual(MapperS.of(economicTerms).<TerminationProvision>map("getTerminationProvision", _economicTerms -> _economicTerms.getTerminationProvision()).<EarlyTerminationProvision>map("getEarlyTerminationProvision", terminationProvision -> terminationProvision.getEarlyTerminationProvision()).<OptionalEarlyTermination>map("getOptionalEarlyTermination", earlyTerminationProvision -> earlyTerminationProvision.getOptionalEarlyTermination()).<ExerciseTerms>map("getExerciseTerms", optionalEarlyTermination -> optionalEarlyTermination.getExerciseTerms()).<OptionExerciseStyleEnum>map("getStyle", exerciseTerms -> exerciseTerms.getStyle()), MapperS.of(OptionExerciseStyleEnum.BERMUDA), CardinalityOperator.All)).get();
			
			return is_Product;
		}
	}
}

File: cdm-sources/src/cdm/margin/schedule/functions/IsIRSwapWithCallableBermudanRightToEnterExitSwaps.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.template.OptionExerciseStyleEnum import OptionExerciseStyleEnum
from cdm.product.template.EconomicTerms import EconomicTerms

__all__ = ['IsIRSwapWithCallableBermudanRightToEnterExitSwaps']


@replaceable
def IsIRSwapWithCallableBermudanRightToEnterExitSwaps(economicTerms: EconomicTerms) -> bool:
    """
    Identifies a product as an IR swap with bermudan/callable right to enter/exit swaps.
    
    Parameters 
    ----------
    economicTerms : EconomicTerms
    
    Returns
    -------
    is_Product : boolean
    
    """
    self = inspect.currentframe()
    
    
    is_Product =  (Qualify_BaseProduct_IRSwap(rosetta_resolve_attr(self, "economicTerms")) and all_elements(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, "economicTerms"), "terminationProvision"), "earlyTerminationProvision"), "optionalEarlyTermination"), "exerciseTerms"), "style"), "=", rosetta_resolve_attr(OptionExerciseStyleEnum, "BERMUDA")))
    
    
    return is_Product

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
